pkg/typesystem/reachabilitygraph.go

```diff
@@ -142,6 +144,78 @@ func (rg *ReachabilityGraph) RelationsEncounteredForResource(
 	return relationRefs, nil
 }
 
+// RelationsEncounteredForSubject returns all relations that are encountered when walking outward from a subject+relation.
+func (rg *ReachabilityGraph) RelationsEncounteredForSubject(
+	ctx context.Context,
+	allDefinitions []*core.NamespaceDefinition,
+	startingSubjectType *core.RelationReference,
+) ([]*core.RelationReference, error) {
+	if startingSubjectType.Namespace != rg.ts.nsDef.Name {
+		return nil, spiceerrors.MustBugf("gave mismatching namespace name for subject type to reachability graph")
+	}
+
+	allRelationNames := mapz.NewSet[string]()
+
+	subjectTypesToCheck := []*core.RelationReference{startingSubjectType}
+
+	// TODO(jschorr): optimize this to not require walking over all types recursively.
+	added := mapz.NewSet[string]()
+	for {
+		if len(subjectTypesToCheck) == 0 {
+			break
+		}
+
+		collected := &[]ReachabilityEntrypoint{}
+		for _, nsDef := range allDefinitions {
+			nts, err := rg.ts.TypeSystemForNamespace(ctx, nsDef.Name)
+			if err != nil {
+				return nil, err
+			}
+
+			nrg := ReachabilityGraphFor(&ValidatedNamespaceTypeSystem{nts})
```

Comment:
nit: shouldn't `TypeSystemForNamespace` return a `ValidatedNamespaceTypeSystem` then?

Reply:
No; we know (internally here) that it is already validated but if the type system was loaded from memory, it may not have been validated yet